File: src/CodeChef/oct18b/ccircles.py
from math import sqrt


def two_point_distance(x1, y1, x2, y2):
    return ((x2 - x1) ** 2 + (y2 - y1) ** 2) ** 0.5


def lies_between(current_circle, small_circle, big_circle):
    center = (small_circle[0], small_circle[1])
    radius_small = small_circle[2]
    radius_big = big_circle[2]
    dist = two_point_distance(center[0], center[1], current_circle[0], current_circle[1])
    nd = dist - current_circle[2]
    fd = dist + current_circle[2]
    if nd <= radius_small and fd >= radius_big:
        return True
    elif radius_small <= nd <= radius_big:
        return True
    elif nd <= radius_small <= fd <= radius_big:
        return True
    else:
        return False


if __name__ == '__main__':
    n, q = [int(__) for __ in input().strip().split()]
    circles = []
    for _ in range(n):
        x, y, r = [int(__) for __ in input().strip().split()]
        circles.append((x, y, r))

    for _ in range(q):
        k = int(input().strip())
        ans = 0
        for i in range(len(circles)):
            for j in range(i + 1, len(circles)):
                x1, y1, r1 = circles[i]
                x2, y2, r2 = circles[j]
                c1c2 = two_point_distance(x1, y1, x2, y2)
                small_circle = (x1, y1, abs(r1 - k))
                big_circle = (x1, y1, r1 + k)
                current_circle = circles[j]
                if lies_between(current_circle, small_circle, big_circle):
                    ans += 1
                # if c1c2 >= r1 + r2:
                #     # Outside
                #     small_circle = (x1, y1, abs(r1 - k))
                #     big_circle = (x1, y1, r1 + k)
                #     current_circle = circles[j]
                #     if lies_between(current_circle, small_circle, big_circle):
                #         ans += 1
                # else:
                #     # Inside
                #
                #     pass
        print(ans)

File: src/CodeChef/oct18b/ccircles.java
package CodeChef.oct18b;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

import static java.lang.Math.abs;
import static java.lang.Math.pow;
import static java.lang.Math.sqrt;

/**
 * Created by bk on 11-10-2018.
 */
public class ccircles {
    public static void main(String[] __) {
        FastIO sc = new FastIO();
        int n = sc.nextInt();
        int q = sc.nextInt();

        ArrayList<Circle> circles = new ArrayList<>();
        for (int i = 0; i < n; i++) circles.add(new Circle(sc.nextLong(), sc.nextLong(), sc.nextLong()));
        for (int q_i = 0; q_i < q; q_i++) {
            int k = sc.nextInt();
            int ans = 0;
            for (int i = 0; i < n; i++) {
                for (int j = i + 1; j < n; j++) {
                    Circle c1 = circles.get(i);
                    Circle c2 = circles.get(j);
                    double c1c2 = twoPointDistance(c1, c2);
                    Circle small = new Circle(c1.x, c1.y, abs(c1.r - k));
                    Circle big = new Circle(c1.x, c1.y, c1.r + k);
                    if (liesBetween(c2, small, big)) ans += 1;
                }
            }
        }
    }

    private static boolean liesBetween(Circle c2, Circle small, Circle big) {
        Point center = new Point(small.x, small.y);
        long radius_small = small.r;
        long radius_big = big.r;
        double dist = twoPointDistance(small, c2);
        double nd = dist - c2.r;
        double fd = dist + c2.r;
        if (nd <= radius_small && fd >= radius_big) return true;
        else if (radius_small <= nd && nd <= radius_big) return true;
        else return nd <= radius_small && radius_small <= fd && fd <= radius_big;
    }

    private static double twoPointDistance(Circle c1, Circle c2) {
        return sqrt(pow(abs(c1.x - c2.x), 2) + pow(abs(c1.y - c2.y), 2));
    }

    static class Point {
        long x;
        long y;

        Point(long x, long y) {
            this.x = x;
            this.y = y;
        }
    }

    static class Circle {
        long x, y, r;

        Circle(long x, long y, long r) {
            this.x = x;
            this.y = y;
            this.r = r;
        }
    }

    static class FastIO {
        private BufferedReader br;
        private StringTokenizer st;

        FastIO() {
            br = new BufferedReader(new
                    InputStreamReader(System.in));
        }

        public String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        public int nextInt() {
            return Integer.parseInt(next());
        }

        public long nextLong() {
            return Long.parseLong(next());
        }

        public double nextDouble() {
            return Double.parseDouble(next());
        }

        public String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }

}
